### PR TITLE
fixed issue where discounts caused wrong params to be send to Mollie …

### PR DESCRIPTION
**Pull request by margud**
…api and breaking it

### CODE REVIEW

File: src/Service/CartLinesService.php [hunk 1]
```diff
@@ -73,17 +74,20 @@ public function getCartLines($amount, $paymentFee, Cart $cart)
         $remaining = round($amount, $apiRoundingPrecision);
         $shipping = round($cart->getTotalShippingCost(null, true), $apiRoundingPrecision);
         $cartSummary = $cart->getSummaryDetails();
-        $cartItems = $cart->getProducts();
+        $cartItems = $cart->getProductsWithSeparatedGifts();
         $wrapping = Configuration::get('PS_GIFT_WRAPPING') ? round($cartSummary['total_wrapping'], $apiRoundingPrecision) : 0;
-        $remaining = round($remaining - $shipping - $wrapping, $apiRoundingPrecision);
-        $totalDiscounts = isset($cartSummary['total_discounts']) ? round($cartSummary['total_discounts'], $apiRoundingPrecision) : 0;
+        $totalDiscounts = $cart->getDiscountSubtotalWithoutGifts();
+        $remaining = round($remaining + $shipping + $wrapping - $totalDiscounts, $apiRoundingPrecision);
```
Comment (JevgenijVisockij) [inline src/Service/CartLinesService.php:81]: I assume those numbers are floats? In PrestaShop its better to use Number class for any sort of math as floats sometimes produce inaccuracies.

https://en.wikipedia.org/wiki/Floating-point_arithmetic#Accuracy_problems

Comment (margud) [reply src/Service/CartLinesService.php:81]: forgot about this one :D

File: src/Service/CartLinesService.php [hunk 2]
```diff
@@ -308,12 +324,12 @@ public static function spreadCartLineGroup($cartLineGroup, $newTotal)
                 'quantity' => $qty,
                 'unitPrice' => (float)$unitPrice,
                 'totalAmount' => (float)$unitPrice * $qty,
-                'sku' => $cartLineGroup[0]['sku'],
+                'sku' => isset($cartLineGroup[0]['sku']) ? $cartLineGroup[0]['sku'] : '',
```
Comment (aleksiuno) [inline src/Service/CartLinesService.php:327]: you can use ternary statement here `'sku' => $cartLineGroup[0]['sku'] ?? '',`

**Review by aleksiuno — APPROVED** [review]

File: src/Utility/NumberUtility.php [hunk 3]
```diff
@@ -89,6 +89,24 @@ public static function toPercentageIncrease($percentage)
...
 
+
+    /**
+     * ($a*$b)
+     *
+     * @param float $a
+     * @param float $b
+     *
+     * @return float
+     */
```
Comment (tomas862) [inline src/Utility/NumberUtility.php:100]: add unit test here maybe ? :)

File: src/Service/CartLinesService.php [hunk 4]
```diff
@@ -73,17 +74,30 @@ public function getCartLines($amount, $paymentFee, Cart $cart)
...
+            NumberUtility::minus(
+                NumberUtility::plus(
+                    NumberUtility::plus($remaining, $shipping),
+                    $wrapping
+                ),
+                $totalDiscounts
+            ),
+            $apiRoundingPrecision
+        );
+        round($remaining + $shipping + $wrapping - $totalDiscounts, $apiRoundingPrecision);
```
Comment (tomas862) [inline src/Service/CartLinesService.php:90]: maybe wrap this thing up and add unit tests as well?

**Pull request merged**